### ISSUE TITLE
Feature Request: Support Seed-Coder

### ISSUE THREAD

**Issue by sainnhe** [ggml-org/llama.cpp#17900]
### Prerequisites

- [x] I am running the latest code. Mention the version if possible as well.
- [x] I carefully followed the [README.md](https://github.com/ggml-org/llama.cpp/blob/master/README.md).
- [x] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggml-org/llama.cpp/discussions), and have a new and useful enhancement to share.

### Feature Description

Bytedance's [seed-coder](https://huggingface.co/ByteDance-Seed/Seed-Coder-8B-Base) supports fim tasks, but currently llama.cpp will return 501 in `/infill`.

The fim prefix and suffix seems to be unsupported. Seed-coder uses `'<[fim-suffix]>'` and `'<[fim-prefix]>'` as mentioned in its readme. It'll be awesome if llama.cpp can support this fim prefix/suffix.

### Motivation

Seed-coder seems to be a very powerful 7b model on handling fim tasks. In their evaluation, it outperforms qwen2.5-coder 7b. It'll be very helpful for users like me who use llama.cpp to self host a LLM for daily code completion tasks.

### Possible Implementation

Add seed-coder's suffix and prefix in `src/llama-vocab.cpp`. It might need some extra tests since the documentation in its readme is not very detailed.

**Comment by CISC** [comment]
One could also just add the FIM tokens to the GGUF metadata, using [GGUF Editor](https://huggingface.co/spaces/CISCai/gguf-editor), [gguf-new-metadata](https://github.com/ggml-org/llama.cpp/blob/master/gguf-py/gguf/scripts/gguf_new_metadata.py) or [gguf-editor-gui](https://github.com/ggml-org/llama.cpp/blob/master/gguf-py/gguf/scripts/gguf_editor_gui.py).

If you visit the `Help` section in `GGUF Editor` it has a guide for just this.

**Comment by sainnhe** [comment]
Thanks for your info! I’ll try it out 👍

**Comment by sainnhe** [comment]
@CISC I just tried GGUF editor, and I bind 3 fim tokens to seed-coder's key. Now the llama-server won't return 501, but I still cannot see any ghost text in my editor.

<img width="1553" height="942" alt="Image" src="https://github.com/user-attachments/assets/6ecf2da4-d1dc-4866-9240-08032fcb03f3" />

Perm link: [Seed-Coder-8B-Base-GGUF/Seed-Coder-8B-Base.Q4_K_M.gguf](https://ciscai-gguf-editor.hf.space/download/mradermacher/Seed-Coder-8B-Base-GGUF/Seed-Coder-8B-Base.Q4_K_M.gguf?add=%5B%22tokenizer.ggml.fim_mid_token_id%22,4,126%5D&add=%5B%22tokenizer.ggml.fim_pre_token_id%22,4,124%5D&add=%5B%22tokenizer.ggml.fim_suf_token_id%22,4,125%5D)

`llama-server` was built from master branch with cuda backend. The editor plugin is llama.vim

Launch command:

```shell
llama-server \
            --host :: \
            --port 8080 \
            -ngl 99 \
            -c 32768 \
            -ctk q8_0 \
            -ctv q8_0 \
            -b 4096 \
            -ub 512 \
            --temp 0.15 \
            --top-k 40 \
            --top-p 0.9 \
            --min-p 0.05 \
            --repeat-penalty 1.0 \
            -fa on \
            --cache-reuse 256 \
            -m ~/.cache/llama.cpp/custom/mradermacher_Seed-Coder-8B-Base-GGUF_Seed-Coder-8B-Base.Q4_K_M.edited.gguf
```

Log:

```log
slot update_slots: id  3 | task 0 | prompt done, n_tokens = 603, batch.n_tokens = 603
slot print_timing: id  3 | task 0 |
prompt eval time =     528.63 ms /   603 tokens (    0.88 ms per token,  1140.68 tokens per second)
       eval time =      94.42 ms /     2 tokens (   47.21 ms per token,    21.18 tokens per second)
      total time =     623.05 ms /   605 tokens
slot      release: id  3 | task 0 | stop processing: n_tokens = 604, truncated = 0
srv  update_slots: all slots are idle
srv  log_server_r: request: POST /infill ::ffff:9.135.95.140 200
slot get_availabl: id  3 | task -1 | selected slot by LCP similarity, sim_best = 0.626 (> 0.100 thold), f_keep = 0.627
slot launch_slot_: id  3 | task -1 | sampler chain: logits -> logit-bias -> top-k -> top-p -> infill -> dist
slot launch_slot_: id  3 | task 3 | processing task
slot update_slots: id  3 | task 3 | new prompt, n_ctx_slot = 32768, n_keep = 0, task.n_tokens = 605
slot update_slots: id  3 | task 3 | n_tokens = 379, memory_seq_rm [379, end)
slot update_slots: id  3 | task 3 | prompt processing progress, n_tokens = 605, batch.n_tokens = 226, progress = 1.000000
slot update_slots: id  3 | task 3 | prompt done, n_tokens = 605, batch.n_tokens = 226
slot print_timing: id  3 | task 3 |
prompt eval time =      10.24 ms /   226 tokens (    0.05 ms per token, 22078.94 tokens per second)
       eval time =     216.34 ms /     2 tokens (  108.17 ms per token,     9.24 tokens per second)
      total time =     226.58 ms /   228 tokens
slot      release: id  3 | task 3 | stop processing: n_tokens = 606, truncated = 0
srv  update_slots: all slots are idle
srv  log_server_r: request: POST /infill ::ffff:9.135.95.140 200
slot get_availabl: id  3 | task -1 | selected slot by LCP similarity, sim_best = 0.582 (> 0.100 thold), f_keep = 0.579
slot launch_slot_: id  3 | task -1 | sampler chain: logits -> logit-bias -> top-k -> top-p -> infill -> dist
slot launch_slot_: id  3 | task 6 | processing task
slot update_slots: id  3 | task 6 | new prompt, n_ctx_slot = 32768, n_keep = 0, task.n_tokens = 603
slot update_slots: id  3 | task 6 | n_tokens = 351, memory_seq_rm [351, end)
slot update_slots: id  3 | task 6 | prompt processing progress, n_tokens = 603, batch.n_tokens = 252, progress = 1.000000
slot update_slots: id  3 | task 6 | prompt done, n_tokens = 603, batch.n_tokens = 252
slot print_timing: id  3 | task 6 |
prompt eval time =      15.50 ms /   252 tokens (    0.06 ms per token, 16261.21 tokens per second)
       eval time =       0.00 ms /     1 tokens (    0.00 ms per token, 1000000.00 tokens per second)
      total time =      15.50 ms /   253 tokens
slot      release: id  3 | task 6 | stop processing: n_tokens = 603, truncated = 0
srv  update_slots: all slots are idle
srv  log_server_r: request: POST /infill ::ffff:9.135.95.140 200
slot get_availabl: id  3 | task -1 | selected slot by LCP similarity, sim_best = 0.581 (> 0.100 thold), f_keep = 0.582
slot launch_slot_: id  3 | task -1 | sampler chain: logits -> logit-bias -> top-k -> top-p -> infill -> dist
slot launch_slot_: id  3 | task 8 | processing task
slot update_slots: id  3 | task 8 | new prompt, n_ctx_slot = 32768, n_keep = 0, task.n_tokens = 604
slot update_slots: id  3 | task 8 | n_tokens = 351, memory_seq_rm [351, end)
slot update_slots: id  3 | task 8 | prompt processing progress, n_tokens = 604, batch.n_tokens = 253, progress = 1.000000
slot update_slots: id  3 | task 8 | prompt done, n_tokens = 604, batch.n_tokens = 253
slot print_timing: id  3 | task 8 |
prompt eval time =      10.47 ms /   253 tokens (    0.04 ms per token, 24168.90 tokens per second)
       eval time =     219.33 ms /     2 tokens (  109.66 ms per token,     9.12 tokens per second)
      total time =     229.80 ms /   255 tokens
slot      release: id  3 | task 8 | stop processing: n_tokens = 605, truncated = 0
srv  update_slots: all slots are idle
srv  log_server_r: request: POST /infill ::ffff:9.135.95.140 200
slot get_availabl: id  3 | task -1 | selected slot by LCP similarity, sim_best = 0.581 (> 0.100 thold), f_keep = 0.580
slot launch_slot_: id  3 | task -1 | sampler chain: logits -> logit-bias -> top-k -> top-p -> infill -> dist
slot launch_slot_: id  3 | task 11 | processing task
slot update_slots: id  3 | task 11 | new prompt, n_ctx_slot = 32768, n_keep = 0, task.n_tokens = 604
slot update_slots: id  3 | task 11 | n_tokens = 351, memory_seq_rm [351, end)
slot update_slots: id  3 | task 11 | prompt processing progress, n_tokens = 604, batch.n_tokens = 253, progress = 1.000000
slot update_slots: id  3 | task 11 | prompt done, n_tokens = 604, batch.n_tokens = 253
slot print_timing: id  3 | task 11 |
prompt eval time =      10.60 ms /   253 tokens (    0.04 ms per token, 23865.67 tokens per second)
       eval time =     219.03 ms /     2 tokens (  109.52 ms per token,     9.13 tokens per second)
      total time =     229.63 ms /   255 tokens
slot      release: id  3 | task 11 | stop processing: n_tokens = 605, truncated = 0
srv  update_slots: all slots are idle
srv  log_server_r: request: POST /infill ::ffff:9.135.95.140 200
```

**Comment by CISC** [comment]
According to the logs it's only returning 1 or 2 tokens, and looking at the [official example](https://huggingface.co/ByteDance-Seed/Seed-Coder-8B-Base#fill-in-the-middle-fim-example) it seems you have to enable SPM infill for it to work correctly; `--spm-infill`.

**Comment by sainnhe** [comment]
It works nicely after adding `--spm-infill`. Thank you so much!
